File: Python/shortestBridge.py
#!/usr/bin/env python3
# encoding: utf-8

"""
@Filename: shortestBridge.py
@Function: 最短的桥 广度优先搜索
@Link: https://leetcode-cn.com/problems/shortest-bridge/
@Python Version: 3.8
@Author: Wei Li
@Date：2021-07-11
"""

import collections


class Solution(object):
    def shortestBridge(self, A):
        R, C = len(A), len(A[0])

        def neighbors(r, c):
            for nr, nc in ((r-1, c), (r, c-1), (r+1, c), (r, c+1)):
                if 0 <= nr < R and 0 <= nc < C:
                    yield nr, nc

        def get_components():
            done = set()
            components = []
            for r, row in enumerate(A):
                for c, val in enumerate(row):
                    if val and (r, c) not in done:
                        # Start dfs
                        stack = [(r, c)]
                        seen = {(r, c)}
                        while stack:
                            node = stack.pop()
                            for nei in neighbors(*node):
                                if A[nei[0]][nei[1]] and nei not in seen:
                                    stack.append(nei)
                                    seen.add(nei)
                        done |= seen
                        components.append(seen)
            return components

        source, target = get_components()
        # print(source, target)
        queue = collections.deque([(node, 0) for node in source])
        done = set(source)
        while queue:
            node, d = queue.popleft()
            if node in target:
                return d-1
            for nei in neighbors(*node):
                if nei not in done:
                    queue.append((nei, d+1))
                    done.add(nei)


# -------------------------
if __name__ == "__main__":
    # A = [[0, 1], [1, 0]]
    A = [[0, 1, 0], [0, 0, 0], [0, 0, 1]]
    # A = [[1, 1, 1, 1, 1], [1, 0, 0, 0, 1], [
    #     1, 0, 1, 0, 1], [1, 0, 0, 0, 1], [1, 1, 1, 1, 1]]

    solution = Solution()
    min_bridge = solution.shortestBridge(A)
    print(f"The solutio of this problem is : {min_bridge}")

File: C++/shortestBridge.cpp
/**
 * @File    : shortestBridge.cpp
 * @Brief   :  最短的桥 广度优先搜索
 * @Link    : https://leetcode-cn.com/problems/shortest-bridge/
 * @Author  : Wei Li
 * @Date    : 2021-07-11
*/

#include <iostream>
#include <vector>
#include <queue>
#include <utility>

class Solution
{
    std::vector<int> direction{-1, 0, 1, 0, -1};

public:
    // 主函数
    int shortestBridge(std::vector<std::vector<int>> &grid)
    {
        int m = grid.size(), n = grid[0].size();
        std::queue<std::pair<int, int>> points;
        // dfs寻找第一个岛屿， 并把1全部赋值为2
        bool flipped = false;
        for (int i = 0; i < m; ++i)
        {
            if (flipped)
                break;
            for (int j = 0; j < n; ++j)
            {
                if (grid[i][j] == 1)
                {
                    dfs(points, grid, m, n, i, j);
                    flipped = true;
                    break;
                }
            }
        }
        // bfs寻找第二个岛屿， 并把过程中经过的0赋值为2
        int x, y;
        int level = 0;
        while (!points.empty())
        {
            ++level;
            int n_points = points.size();
            while (n_points--)
            {
                // auto [r, c] = points.front(); // c++17
                auto r = points.front().first;
                auto c = points.front().second;
                points.pop();
                for (int k = 0; k < 4; ++k)
                {
                    x = r + direction[k], y = c + direction[k + 1];
                    if (x >= 0 && y >= 0 && x < m && y < n)
                    {
                        if (grid[x][y] == 2)
                        {
                            continue;
                        }
                        if (grid[x][y] == 1)
                        {
                            return level;
                        }
                        points.push({x, y});
                        grid[x][y] = 2;
                    }
                }
            }
        }
        return 0;
    }
    // 辅函数
    void dfs(std::queue<std::pair<int, int>> &points, std::vector<std::vector<int>> &grid, int m, int n, int i, int j)
    {
        if (i < 0 || j < 0 || i == m || j == n || grid[i][j] == 2)
        {
            return;
        }
        if (grid[i][j] == 0)
        {
            points.push({i, j});
            return;
        }
        grid[i][j] = 2;
        dfs(points, grid, m, n, i - 1, j);
        dfs(points, grid, m, n, i + 1, j);
        dfs(points, grid, m, n, i, j - 1);
        dfs(points, grid, m, n, i, j + 1);
    }
};

// ---------------------------------
int main(int argc, char **argv)
{
    // std::vector<std::vector<int>> grid{{0, 1}, {1, 0}};
    // std::vector<std::vector<int>> grid{{0, 1, 0}, {0, 0, 0}, {0, 0, 1}};
    std::vector<std::vector<int>> grid{{1, 1, 1, 1, 1}, {1, 0, 0, 0, 1}, {1, 0, 1, 0, 1}, {1, 0, 0, 0, 1}, {1, 1, 1, 1, 1}};

    Solution solution;
    auto min_bridge = solution.shortestBridge(grid);
    std::cout << "The solutio of this problem i s: " << min_bridge << std::endl;

    return 0;
}
